airbyte-integrations/connectors/source-s3/source_s3/v4/unstructured_parser.py

```diff
@@ -0,0 +1,94 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+import logging
+from io import IOBase
+from typing import Any, Dict, Iterable, List, Mapping, Optional
+
+from airbyte_cdk.sources.file_based.config.file_based_stream_config import FileBasedStreamConfig
+from airbyte_cdk.sources.file_based.file_based_stream_reader import AbstractFileBasedStreamReader, FileReadMode
+from airbyte_cdk.sources.file_based.file_types.file_type_parser import FileTypeParser
+from airbyte_cdk.sources.file_based.remote_file import RemoteFile
+from airbyte_cdk.sources.file_based.schema_helpers import SchemaType
+from source_s3.v4.config import S3FileBasedStreamConfig
+
+
+class UnstructuredParser(FileTypeParser):
+
+    MAX_SIZE_PER_CHUNK = 4_000_000
+
+    async def infer_schema(
+        self,
+        config: S3FileBasedStreamConfig,
+        file: RemoteFile,
+        stream_reader: AbstractFileBasedStreamReader,
+        logger: logging.Logger,
+    ) -> SchemaType:
+        return {
+            "content": {"type": "string"},
+            "chunk_number": {"type": "integer"},
+            "no_of_chunks": {"type": "integer"},
+            "id": {"type": "string"},
+        }
+
+    def parse_records(
+        self,
+        config: FileBasedStreamConfig,
+        file: RemoteFile,
+        stream_reader: AbstractFileBasedStreamReader,
+        logger: logging.Logger,
+        discovered_schema: Optional[Mapping[str, SchemaType]],
+    ) -> Iterable[Dict[str, Any]]:
+        with stream_reader.open_file(file, self.file_read_mode, None, logger) as fp:
+            markdown = self.read_file(fp, logger)
+            if not markdown:
+                return []
+            chunks = [markdown[i : i + self.MAX_SIZE_PER_CHUNK] for i in range(0, len(markdown), self.MAX_SIZE_PER_CHUNK)]
+            yield from [
+                {
+                    "content": chunk,
+                    "chunk_number": i,
+                    "no_of_chunks": len(chunks),
+                    "id": f"{file.uri}_{i}",
+                }
+                for i, chunk in enumerate(chunks)
+            ]
```

Comment:
Feels like this should yield chunks instead of yielding from the list.

```suggestion
            chunk_number = 0
            no_of_chunks = ceil(len(markdown) / self.MAX_SIZE_PER_CHUNK)
            for chunk in chunked(markdown, self.MAX_SIZE_PER_CHUNK):
                yield  {
                    "content": chunk,
                    "chunk_number": chunk_number,
                    "no_of_chunks": no_of_chunks,
                    "id": f"{file.uri}_{i}",
                }
                chunk_number += 1
```

Reply:
Good catch, @clnoll. The list comprehension causes all items to be collected before any are yielded.

Little-known trick, there's also a less-used [generator comprehension](https://www.pythonlikeyoumeanit.com/Module2_EssentialsOfPython/Generators_and_Comprehensions.html#Creating-your-own-generator:-generator-comprehensions) - with almost identical syntax.

```suggestion
            chunks = (markdown[i : i + self.MAX_SIZE_PER_CHUNK] for i in range(0, len(markdown), self.MAX_SIZE_PER_CHUNK))
            yield from (
                {
                    "content": chunk,
                    "chunk_number": i,
                    "no_of_chunks": len(chunks),
                    "id": f"{file.uri}_{i}",
                }
                for i, chunk in enumerate(chunks)
            )
```

Same inline syntax as the original, but should perform similarly to @clnoll's non-blocking proposal, which emits records chunks one at a time.

Reply:
As we decided to go without chunks for now, this got removed - still thanks for bringing it up, I didn't know about the generator comprehensions